### ISSUE TITLE
Proposal: Translation Support for the Try Scheme Tutorial and User Interface

### ISSUE THREAD

**Issue by nattybear** [gambit/gambit#934]
Hello Gambit Team,

I am interested in contributing by translating the "Start Tutorial" from try.scheme.org into Korean. I would also like to help translate the rest of the user interface.

Based on [this discussion](https://github.com/schemeorg-community/try.scheme.org/issues/8#issuecomment-2396841891), it seems that having a static HTML document and a preprocessor for runnable examples would be the best approach. Please let me know if this is the right direction or if you have other suggestions.

Thank you for your consideration.

**Comment by feeley** [comment]
That would be a great contribution!

To get you started you should extract the HTML documentation by executing this at the REPL

```
(for-each
  (lambda (topic)
    (write-file-string (string-append "tutorial_" topic ".html")
                       ((host-eval "(topic) => main_vm.ui.tutorial_sections[topic]") topic)))
  '("basic" "types" "forms" "libraries" "keybindings"))
```

This will create the files `tutorial_basic.html`, etc which you can then view with the `+` tab of the file panel and selecting the file to view. By default this will let the browser display the HTML in the file. You can click on the file's header to select "Toggle text and HTML views" so that you can see (and possibly edit) the HTML markup. It is probably best to copy these files to your preferred editor to do the translation there (or ask chatgpt to do the translation!).

You could then replace the strings in `main_vm.ui.tutorial_sections['<topic>']` with your translations to see if they display properly. I will start thinking of a way to detect and/or select the user's preferred language.

**Comment by nattybear** [comment]
As a first test, I translated only `tutorial_basic.html` and saved it as `tutorial_basic_ko.html`. I have confirmed that it displays correctly. Thank you for showing me the method.

However, I would like to suggest using the GitHub repository instead of directly uploading to the online environment.

It might also be a good idea to enable the use of Markdown rather than directly editing HTML.

**Comment by feeley** [comment]
> However, I would like to suggest using the GitHub repository instead of directly uploading to the online environment.
> 
> It might also be a good idea to enable the use of Markdown rather than directly editing HTML.

I'm not sure what you are suggesting. Do you mean adding the tutorial to the repository on github? This would not allow the examples to be run using a click on the "run" button. For a tight integration of the tutorial the text of the tutorial is in the web app, in HTML format. There could be an external preprocessor that converts markdown to HTML, but this is nontrivial.

You might want to coordinate with @maconcg that is rewriting the Gambit documentation (see #936).

**Comment by maconcg** [comment]
@nattybear If you're able, try cloning my branch that's linked from #936.  You can edit the file `doc/rrg/toolshed/color-samples.txi` and then, from the `doc/rrg` directory, run `make sample`.  It'll output a webpage like this one:

[gambit-color-samples.html.gz](https://github.com/user-attachments/files/17348250/gambit-color-samples.html.gz)

It's not Markdown, but I'm hoping people will find it's not much harder to work with.  You can put code examples between `@lisp` and `@end lisp` -- the only special markup you need to use within `@lisp` block is:

- If an expression returns an exception, wrap it in `@exception{...}`
- Otherwise, wrap the return value for an exception in `@ok{...}`.

This is probably the easiest way to try it out.

A real example of a file containing procedure descriptions is `doc/rrg/procedures/pairs-lists.txi`.  Some examples of “regular” prose are `doc/rrg/prose/intro.txi` and `doc/rrg/prose/pairs-lists.txi`.  You would only need to edit files that look these three.  Do you think this is a format you could work with?

Also, when you mention using the Github repository, do you mean Github Pages?

**Comment by nattybear** [comment]
What I meant by mentioning the GitHub repository was the following:

* Contributors upload the translated documents to the GitHub repository.
* The Gambit online environment fetches and renders the files stored in the GitHub repository.

I believe this approach would allow multiple people to contribute to the documentation transparently.

I understand that handling Markdown is not a trivial issue. It would be best to discuss this matter separately at a later time.

**Comment by nattybear** [comment]
I have reviewed the output generated using Texinfo as you suggested. While I still need to study Texinfo further, it seems like a useful tool.

However, I’m concerned that even with Texinfo, we may not be able to implement the "run" button, as feeley mentioned.

The method using Texinfo seems suitable for working on Gambit documentation, but I’m unsure if it will help resolve this issue. (I am interested in both this issue and the overall Gambit documentation work.)

Meanwhile, I took a look at codeboot.org, which [feeley mentioned](https://github.com/schemeorg-community/try.scheme.org/issues/8#issuecomment-2396841891), and it seems quite similar to the Gambit online environment. I was also impressed by how it provides services in multiple languages.

**Comment by maconcg** [comment]
Right, Texinfo won't be the _whole_ solution.  Basically, it gives us a standard storage format and syntax-highlighted static HTML.  As currently implemented, the static-HTML syntax highlighting works by preprocessing the Texinfo source and then feeding that processed source to `texi2any`.  There's also a basic extraction script at `doc/rrg/toolshed/extract-tests.sh`, which outputs one file per code example.

For runnable examples, we'd probably go with one of these options:

1. Insert dynamic elements into the Texinfo-generated HTML.
2. Extract the prose and syntax-highlighted code examples from the Texinfo-generated HTML and insert that content into a new page that includes the web IDE/run button.

**Comment by nattybear** [comment]
Using Texinfo for syntax highlighting and similar features looks great and seems very helpful for Gambit documentation.

However, I’m not sure how much it will help with this Try Gambit translation issue.

For this issue, there’s actually no need to write new code. It’s simply a matter of rewriting the English prose into Korean.

[Here is a gist link](https://gist.github.com/nattybear/739d71fdf7db38d16080a1cfd9bc5a00) where I’ve translated a portion of it with the help of ChatGPT.
